### PR TITLE
catalog: Move mz_adapter::catalog to mz_catalog

### PR DESCRIPTION
**Pull request by jkosh44**
This commit moves some files out of the mz_adapter::catalog module and into the mz_catalog crate. It does so by creating an mz_adapter-types crate to hold common types.

Works towards resolving #22593

### Motivation
This PR refactors existing code.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

File: src/catalog/src/durable.rs
Comment (jkosh44) [inline src/catalog/src/durable.rs:1]: I'm not sure why this whole file is marked as new, but it's pretty much the same as the old `src/catalog/src/lib.rs` with some changed imports and module declarations.

**Comment by jkosh44** [comment]
`mz-adapter-types` is a bit empty, but hopefully we can fill it with more as time goes on.

**Comment by jkosh44** [comment]
@danhhz I'm curious if you approve of the high level attempt to de-tangle the catalog and adapter, since you're becoming an expert in dependency management.

**Review by danhhz — COMMENTED** [review]
Yup, seems sane! Doesn't look like it's gonna get you much change in the dep graph, but the finer grained invalidation might help speed things up a bit when iterating on mz-adapter. Here's the new graph btw in the case the picture is helpful for anyone

image

File: src/catalog/src/durable.rs
Comment (jkosh44) [reply src/catalog/src/durable.rs:1]: Done. That seems to work pretty well, but in the GH UI you have to specifically view only the second commit.

**Review by nrainer-materialize — APPROVED** [review]
The [nightly build](https://buildkite.com/materialize/nightlies/builds/4844) can be considered ok (the failures are expected).

**Comment by jkosh44** [comment]
> I mostly just skimmed this, let me know if there is anything I need to look at more closely, but overall LGTM!

It's all code movement, I just wanted to make sure the overall structure made sense and was a good approach.

**Pull request merged**